File: app/src/main/java/org/breezyweather/weather/WeatherHelper.java
package org.breezyweather.weather;

import android.content.Context;

import androidx.annotation.NonNull;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import org.breezyweather.common.basic.models.Location;
import org.breezyweather.common.basic.models.options.provider.WeatherSource;
import org.breezyweather.common.basic.models.weather.Weather;
import org.breezyweather.db.repositories.HistoryEntityRepository;
import org.breezyweather.db.repositories.WeatherEntityRepository;
import org.breezyweather.main.utils.RequestErrorType;
import org.breezyweather.weather.services.WeatherService;
import org.breezyweather.common.rxjava.BaseObserver;
import org.breezyweather.common.rxjava.ObserverContainer;
import org.breezyweather.common.rxjava.SchedulerTransformer;
import org.breezyweather.common.utils.NetworkUtils;
import org.breezyweather.common.utils.helpers.AsyncHelper;

public class WeatherHelper {

    private final WeatherServiceSet mServiceSet;
    private final CompositeDisposable mCompositeDisposable;

    public interface OnRequestWeatherListener {
        void requestWeatherSuccess(@NonNull Location requestLocation);
        void requestWeatherFailed(@NonNull Location requestLocation, RequestErrorType requestErrorType);
    }

    public interface OnRequestLocationListener {
        void requestLocationSuccess(String query, List<Location> locationList);
        void requestLocationFailed(String query, RequestErrorType requestErrorType);
    }

    @Inject
    public WeatherHelper(WeatherServiceSet weatherServiceSet,
                         CompositeDisposable compositeDisposable) {
        mServiceSet = weatherServiceSet;
        mCompositeDisposable = compositeDisposable;
    }

    public void requestWeather(Context c, Location location, @NonNull final OnRequestWeatherListener l) {
        final WeatherService service = mServiceSet.get(location.getWeatherSource());
        if (!NetworkUtils.isAvailable(c)) {
            l.requestWeatherFailed(location, RequestErrorType.NETWORK_UNAVAILABLE);
            return;
        } else if (!location.isUsable()) {
            l.requestWeatherFailed(location, RequestErrorType.LOCATION_FAILED);
            return;
        }

        service.requestWeather(c, location.copy(), new WeatherService.RequestWeatherCallback() {

            @Override
            public void requestWeatherSuccess(@NonNull Location requestLocation) {
                Weather weather = requestLocation.getWeather();
                if (weather != null) {
                    WeatherEntityRepository.INSTANCE.writeWeather(requestLocation, weather);
                    if (weather.getYesterday() == null) {
                        weather.setYesterday(
                                HistoryEntityRepository.INSTANCE.readHistory(requestLocation, weather)
                        );
                    }
                    l.requestWeatherSuccess(requestLocation);
                } else {
                    requestWeatherFailed(requestLocation, RequestErrorType.WEATHER_REQ_FAILED);
                }
            }

            @Override
            public void requestWeatherFailed(@NonNull Location requestLocation, @NonNull RequestErrorType requestErrorType) {
                l.requestWeatherFailed(
                        Location.copy(
                                requestLocation,
                                WeatherEntityRepository.INSTANCE.readWeather(requestLocation)
                        ),
                        requestErrorType
                );
            }
        });
    }

    public void requestSearchLocations(Context context, String query, WeatherSource enabledSource,
                                       @NonNull final OnRequestLocationListener l) {
        if (enabledSource == null) {
            AsyncHelper.delayRunOnUI(() -> l.requestLocationFailed(query, RequestErrorType.LOCATION_SEARCH_FAILED), 0);
            return;
        }

        // generate weather services.
        final WeatherService service = mServiceSet.get(enabledSource);

        if (!service.isConfigured(context)) {
            AsyncHelper.delayRunOnUI(() -> l.requestLocationFailed(query, RequestErrorType.API_KEY_REQUIRED_MISSING), 0);
            return;
        }

        // generate observable list.
        Observable<List<Location>> observable =
                    Observable.create(emitter ->
                            emitter.onNext(service.requestLocation(context, query))
                    );

        observable.compose(SchedulerTransformer.create())
                .subscribe(new ObserverContainer<>(mCompositeDisposable, new BaseObserver<List<Location>>() {
                    @Override
                    public void onSucceed(List<Location> locationList) {
                        if (locationList != null && locationList.size() != 0) {
                            l.requestLocationSuccess(query, locationList);
                        } else {
                            onFailed();
                        }
                    }

                    @Override
                    public void onFailed() {
                        if (this.isApiLimitReached()) {
                            l.requestLocationFailed(query, RequestErrorType.API_LIMIT_REACHED);
                        } else if (this.isApiUnauthorized()) {
                            l.requestLocationFailed(query, RequestErrorType.API_UNAUTHORIZED);
                        } else {
                            l.requestLocationFailed(query, RequestErrorType.LOCATION_SEARCH_FAILED);
                        }
                    }
                }));
    }

    public void cancel() {
        for (WeatherService s : mServiceSet.getAll()) {
            s.cancel();
        }
        mCompositeDisposable.clear();
    }
}

File: app/src/main/java/org/breezyweather/weather/WeatherHelper.kt
package org.breezyweather.weather

import android.content.Context
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.CompositeDisposable
import kotlinx.coroutines.rx3.awaitSingle
import org.breezyweather.R
import org.breezyweather.common.basic.models.Location
import org.breezyweather.common.basic.models.options.provider.WeatherSource
import org.breezyweather.common.basic.models.weather.Weather
import org.breezyweather.common.exceptions.LocationException
import org.breezyweather.common.exceptions.NoNetworkException
import org.breezyweather.common.exceptions.ParsingException
import org.breezyweather.common.extensions.isOnline
import org.breezyweather.common.rxjava.ApiObserver
import org.breezyweather.common.rxjava.ObserverContainer
import org.breezyweather.common.rxjava.SchedulerTransformer
import org.breezyweather.common.utils.helpers.AsyncHelper
import org.breezyweather.db.repositories.HistoryEntityRepository
import org.breezyweather.db.repositories.WeatherEntityRepository
import org.breezyweather.main.utils.RequestErrorType
import javax.inject.Inject

class WeatherHelper @Inject constructor(
    private val mServiceSet: WeatherServiceSet,
    private val mCompositeDisposable: CompositeDisposable
) {
    // TODO: To be deprecated
    interface OnRequestSearchLocationsListener {
        fun requestLocationSuccess(query: String, locationList: List<Location>)
        fun requestLocationFailed(query: String, requestErrorType: RequestErrorType)
    }

    suspend fun getWeather(context: Context, location: Location): Weather {
        return requestWeather(context, location).awaitSingle()
    }

    fun requestWeather(
        context: Context, location: Location
    ): Observable<Weather> {
        val service = mServiceSet[location.weatherSource]
        if (!context.isOnline()) {
            return Observable.error(NoNetworkException())
        } else if (!location.isUsable) {
            return Observable.error(LocationException())
        }

        return service
            .requestWeather(context, location.copy())
            .map { t ->
                if (t.result != null) {
                    WeatherEntityRepository.writeWeather(location, t.result)
                    if (t.result.yesterday == null) {
                        t.result.yesterday = HistoryEntityRepository.readHistory(location, t.result)
                    }
                    t.result
                } else {
                    throw ParsingException()
                }
            }
    }

    fun requestSearchLocations(
        context: Context,
        query: String,
        enabledSource: WeatherSource?,
        listener: OnRequestSearchLocationsListener
    ) {
        if (enabledSource == null) {
            AsyncHelper.delayRunOnUI({ listener.requestLocationFailed(query, RequestErrorType.LOCATION_SEARCH_FAILED) }, 0)
            return
        }

        // generate weather services.
        val service = mServiceSet[enabledSource]
        if (!service.isConfigured(context)) {
            AsyncHelper.delayRunOnUI({ listener.requestLocationFailed(query, RequestErrorType.API_KEY_REQUIRED_MISSING) }, 0)
            return
        }

        // generate observable list.
        Observable.create { emitter ->
            emitter.onNext(service.requestLocationSearch(context, query))
        }.compose(SchedulerTransformer.create())
            .subscribe(ObserverContainer(mCompositeDisposable, object : ApiObserver<List<Location>>() {
                override fun onSucceed(t: List<Location>) {
                    listener.requestLocationSuccess(query, t)
                }

                override fun onFailed() {
                    if (isApiLimitReached) {
                        listener.requestLocationFailed(query, RequestErrorType.API_LIMIT_REACHED)
                    } else if (isApiUnauthorized) {
                        listener.requestLocationFailed(query, RequestErrorType.API_UNAUTHORIZED)
                    } else {
                        listener.requestLocationFailed(query, RequestErrorType.LOCATION_SEARCH_FAILED)
                    }
                }
            }))
    }

    fun cancel() {
        mServiceSet.all.forEach { it.cancel() }
        mCompositeDisposable.clear()
    }
}
